xsnap/sources/xsnapPlatform.c

```diff
@@ -310,69 +347,118 @@ void fxRunDebugger(txMachine* the)
 #endif
 }
 
+/* SHARED TIMERS */
+
+typedef struct sxSharedTimers txSharedTimers;
+struct sxSharedTimers {
+	txSharedTimer* first;
+	txMutex mutex;
+};
+static txSharedTimers gxSharedTimers;
+
+void fxInitializeSharedTimers()
+{
+	c_memset(&gxSharedTimers, 0, sizeof(txSharedTimers));
+	fxCreateMutex(&(gxSharedTimers.mutex));
+}
+
+void fxTerminateSharedTimers()
+{
+	fxDeleteMutex(&(gxSharedTimers.mutex));
+	if (gxSharedTimers.first != C_NULL) {
+		fprintf(stderr, "# shared timers mismatch!\n");
+		exit(1);
+	}
+}
+
+void fxRescheduleSharedTimer(txSharedTimer* timer, txNumber timeout, txNumber interval)
+{
+		fxLockMutex(&(gxSharedTimers.mutex));
+	timer->when = mxMonotonicNow() + timeout;
+	timer->interval = interval;
+		fxUnlockMutex(&(gxSharedTimers.mutex));
+}
+
+void* fxScheduleSharedTimer(txNumber timeout, txNumber interval, txSharedTimerCallback callback, void* refcon, txInteger refconSize)
+{
+	txSharedTimer* timer;
+	txSharedTimer** address;
+	txSharedTimer* link;
+	timer = c_calloc(1, sizeof(txSharedTimer) + refconSize - 1);
+	if (timer) {
+		timer->thread = mxCurrentThread();
+		timer->when = mxMonotonicNow() + timeout;
+		timer->interval = interval;
+		timer->callback = callback;
+		timer->refconSize = refconSize;
+		c_memcpy(timer->refcon, refcon, refconSize);
+
+		fxLockMutex(&(gxSharedTimers.mutex));
+		address = (txSharedTimer**)&(gxSharedTimers.first);
+		while ((link = *address))
+			address = &(link->next);
+		*address = timer;
+		fxUnlockMutex(&(gxSharedTimers.mutex));
+		}
+		return timer;
+}
+
+void fxUnscheduleSharedTimer(txSharedTimer* timer)
+{
+	txSharedTimer** address;
+	txSharedTimer* link;
+		fxLockMutex(&(gxSharedTimers.mutex));
+		address = (txSharedTimer**)&(gxSharedTimers.first);
+	while ((link = *address)) {
+		if (link == timer) {
+			*address = link->next;
+			c_free(timer);
+			break;
+		}
+		address = &(link->next);
+	}
+		fxUnlockMutex(&(gxSharedTimers.mutex));
+}
+
 void fxRunLoop(txMachine* the)
 {
-	c_timeval tv;
+	txThread thread = mxCurrentThread();
 	txNumber when;
-	txJob* job;
-	txJob** address;
+	txInteger count;
+	txSharedTimer* timer;
 	for (;;) {
 		while (the->promiseJobs) {
 			the->promiseJobs = 0;
 			fxRunPromiseJobs(the);
 		}
-		fxEndJob(the);
+    // TODO reason thru this removal projected from Moddable version
```

Comment:
👍 (same), and the other run loop changes below.
Slightly concerned to see mutexes show up around these parts. In general we don't do time in xsnap, and the only thing we're interested in is a 0 timeout / setImmediate, so I wouldn't mind blowing it all to smudges.